### PR TITLE
server: add peer label level statistics

### PR DESCRIPTION
**Pull request by nolouch**
record peer label Isolation level in that store.

### CODE REVIEW

**Comment by disksing** [comment]
I think you can add a new method to store that calculate the minimal location level that each of the region's peers are located on different labels.

**Comment by nolouch** [comment]
PTAL @disksing @Connor1996

File: server/metrics.go [hunk 1]
```diff
@@ -66,6 +66,15 @@ var (
...
 
+	regionLabelLevelHistogram = prometheus.NewHistogram(
+		prometheus.HistogramOpts{
+			Namespace: "pd",
+			Subsystem: "regions",
+			Name:      "label_level",
+			Help:      "Bucketed histogram of the label level of the region.",
+			Buckets:   prometheus.LinearBuckets(0, 1, 8),
+		})
+
```
Comment (disksing) [inline server/metrics.go]: not registered?

File: server/region_statistics.go [hunk 2]
```diff
@@ -104,6 +104,58 @@ func (r *regionStatistics) Observe(region *core.RegionInfo, stores []*core.Store
...
+	regionLabelLevel := getRegionLabelIsolationLevel(stores, labels)
+	regionLabelLevelHistogram.Observe(float64(regionLabelLevel))
+}
+
+func getRegionLabelIsolationLevel(stores []*core.StoreInfo, labels []string) int {
+	if len(stores) == 0 || len(labels) == 0 {
+		return -1
+	}
+	var queueStores [][]*core.StoreInfo
+	queueStores = append(queueStores, stores)
```
Comment (disksing) [inline server/region_statistics.go]: queueStores := [][]*core.StoreInfo{stores}

File: server/region_statistics.go [hunk 3]
```diff
@@ -104,6 +104,58 @@ func (r *regionStatistics) Observe(region *core.RegionInfo, stores []*core.Store
...
+		labelValue := s.GetLabelValue(label)
+		if labelValue == "" {
+			continue
+		}
+		groupStores, ok := m[labelValue]
+		if !ok {
+			groupStores = make([]*core.StoreInfo, 0, defaultMaxReplicas)
+		}
+		groupStores = append(groupStores, s)
+		m[labelValue] = groupStores
```
Comment (disksing) [inline server/region_statistics.go]: I think we can `m[labelValue] = append(m[labelValue], s)` to make it simple.

File: server/region_statistics.go [hunk 4]
```diff
@@ -104,6 +104,58 @@ func (r *regionStatistics) Observe(region *core.RegionInfo, stores []*core.Store
...
+		}
+		queueStores = queueStores[dequeNumber:]
+		if isInLabelIsolation {
+			return level
+		}
+	}
+	return -1
+}
+
+func higherLabelLeverStores(stores []*core.StoreInfo, label string) [][]*core.StoreInfo {
```
Comment (disksing) [inline server/region_statistics.go]: Lever -> Level. And the function name looks ambiguous, maybe we can figure out something better.

File: server/region_statistics.go [hunk 5]
```diff
@@ -104,6 +104,50 @@ func (r *regionStatistics) Observe(region *core.RegionInfo, stores []*core.Store
...
+	if len(stores) == 0 {
+		return
+	}
+	regionLabelLevel := getRegionLabelIsolationLevel(stores, labels)
+	regionLabelLevelHistogram.Observe(float64(regionLabelLevel))
+}
+
+func getRegionLabelIsolationLevel(stores []*core.StoreInfo, labels []string) int {
+	if len(stores) == 0 || len(labels) == 0 {
+		return -1
```
Comment (disksing) [inline server/region_statistics.go]: Seems return 0 is more suitable.

**Pull request merged**